internal/jobobject/iocp.go

```diff
@@ -0,0 +1,67 @@
+package jobobject
+
+import (
+	"sync"
+	"unsafe"
+
+	"github.com/Microsoft/hcsshim/internal/winapi"
+	"golang.org/x/sys/windows"
+)
+
+var (
+	ioInitOnce sync.Once
+	initIOErr  error
+	// Global iocp handle that will be re-used for every job object
+	ioCompletionPort windows.Handle
+	// Mapping of job handle to channel that that job object owns to be able to receive
+	// notifications.
+	jobMap = make(map[uintptr]chan interface{})
+)
+
+// AllProcessesExited is a type representing a notification that every process in a job has exited.
+type AllProcessesExited struct {
+	Err error
+}
+
+// ioCompletionProcessor processes completed async IOs forever
+func pollIOCP(iocpHandle windows.Handle) {
+	var (
+		overlapped uintptr
+		code       uint32
+		key        uintptr
+	)
+
+	for {
+		err := winapi.GetQueuedCompletionStatus(iocpHandle, &code, &key, (**windows.Overlapped)(unsafe.Pointer(&overlapped)), windows.INFINITE)
+		if ch, ok := jobMap[key]; ok {
+			// Check code and parse out relevant information related to that notification
+			// that we care about. For now all we handle is the message that all processes
+			// in the job have exited.
+			switch code {
+			case winapi.JOB_OBJECT_MSG_ACTIVE_PROCESS_ZERO:
+				// Send the AllProcessesExited notification. We don't remove the map entry here
+				// as for container scenarios for all intents and purposes this is the
+				// end of the containers life, but for possible future usecases for this package
+				// there's nothing stopping more processes from being added to the job and this
+				// being fired once the count reaches zero again. A call to JobObject.Close() will
+				// remove the entry as the handle is invalidated at this point and the job is
+				// finished for good.
+				ch <- AllProcessesExited{
+					Err: err,
```

Comment:
I would assume if `GetQueuedCompletionStatus` failed, then we shouldn't trust its return values to have reasonable values at all. In this case, I think we could either send an error on the channel, or just log something here.

Reply:
I think logging might be a better idea. I was originally checking if the type contained an error and logging in the container methods but might as well log at the site of the error

Reply:
Yes, I think logging is probably better, at least while this is an internal package. I can't think of anything the client would want to do in this case.